cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -23,46 +23,87 @@
  * @param destination The bitmask to write result into
  * @param source Array of source mask pointers. All masks must be of same size
  * @param source_begin_bits Array of offsets into corresponding @p source masks.
  *                          Must be same size as source array
  * @param source_size_bits Number of bits in each mask in @p source
+ * @param count Pointer to valid-bit counter
  */
-template <typename Binop>
+template <int block_size, typename Binop>
 __global__ void offset_bitmask_binop(Binop op,
                                      device_span<bitmask_type> destination,
                                      device_span<bitmask_type const*> source,
                                      device_span<size_type const> source_begin_bits,
-                                     size_type source_size_bits)
+                                     size_type source_size_bits,
+                                     size_type* valid_count_ptr)
 {
-  for (size_type destination_word_index = threadIdx.x + blockIdx.x * blockDim.x;
-       destination_word_index < destination.size();
+  constexpr auto const word_size{detail::size_in_bits<bitmask_type>()};
+
+  auto const tid             = threadIdx.x + blockIdx.x * blockDim.x;
+  auto const first_bit_index = 0;
+  auto const last_bit_index  = source_size_bits - 1;
+  auto const first_word_index{word_index(first_bit_index)};
+  auto const last_word_index{word_index(last_bit_index)};
+
+  size_type thread_valid_count = 0;
+
+  for (size_type destination_word_index = tid; destination_word_index < destination.size();
        destination_word_index += blockDim.x * gridDim.x) {
     bitmask_type destination_word =
       detail::get_mask_offset_word(source[0],
                                    destination_word_index,
                                    source_begin_bits[0],
                                    source_begin_bits[0] + source_size_bits);
     for (size_type i = 1; i < source.size(); i++) {
-      destination_word =
-
-        op(destination_word,
-           detail::get_mask_offset_word(source[i],
-                                        destination_word_index,
-                                        source_begin_bits[i],
-                                        source_begin_bits[i] + source_size_bits));
+      destination_word = op(destination_word,
+                            detail::get_mask_offset_word(source[i],
+                                                         destination_word_index,
+                                                         source_begin_bits[i],
+                                                         source_begin_bits[i] + source_size_bits));
     }
 
     destination[destination_word_index] = destination_word;
+    thread_valid_count += __popc(destination_word);
   }
+
+  // Subtract any slack bits counted from the first and last word
+  // Two threads handle this -- one for first word, one for last
```

Comment:
> Would there be a meaningful loss of performance if this work were serialized and done entirely by thread 0?

Probably not, there are no expensive operations involved thus all will happen within several microseconds I guess


> Would the code be any longer?

Yes, we need to duplicate the same set of variables for both `first` and `last`

I understand your concern about code readability and thread divergence brought by branching. In terms of performance, I think this code will cost "little" against the block reduce and atomic operation right after it. Using one single thread to handle both `first` and `last` words may bring other confusions and will make the code longer. What do you think?

Reply:
See suggestion here. Subjectively I see this snippet as a little less confusing, but it's not any longer because we can eliminate intermediate variables that are currently needed to clarify the heavily branched ternary logic for variable initialization. https://github.com/rapidsai/cudf/pull/9616#discussion_r746068104